### PR TITLE
[wip] Make `RichProgressBar` use `auto_refresh=True`

### LINKED ISSUE #13937: `RichProgressBar` in v1.6 is slower than v1.5
## 🐛 Bug
Similarly to (but independently of) #13179, there's a regression in speed of the rich progress bar between 1.5 and 1.6.

```
time: 2.059369374997914   # v1.5.10
time: 13.186531708983239  # v1.6.0
time: 11.988152374979109  # v1.6.5
time: 12.897805292042904  # master (aefb9ab43f9a8e6704558a346dbae1a00044bb45)
```

### To Reproduce

```python
from time import monotonic
import torch
from torch.utils.data import DataLoader, Dataset
from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.callbacks import RichProgressBar

class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len

class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
        return [optimizer], [lr_scheduler]

def main():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    model = BoringModel()
    trainer = Trainer(
        max_epochs=100,
        enable_model_summary=False,
        enable_checkpointing=False,
        logger=False,
        benchmark=False,  # True by default in 1.6.{0-3}.
        callbacks=RichProgressBar(),
    )
    t0 = monotonic()
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
    print("time:", monotonic() - t0)

if __name__ == "__main__":
    main()
```

### Expected behavior
There should be no regression unless there's a reasonable explanation.

### Environment

- Lightning Component (e.g. Trainer, LightningModule, LightningApp, LightningWork, LightningFlow): n/a
- PyTorch Lightning Version (e.g., 1.5.0): master (aefb9ab43f9a8e6704558a346dbae1a00044bb45), 1.6.5, 1.6.0, 1.5.10
- PyTorch Version (e.g., 1.10): 1.12
- Python version (e.g., 3.9): 3.10
- OS (e.g., Linux): macOS
- CUDA/cuDNN version: n/a
- GPU models and configuration: Apple Silicon M1
- How you installed PyTorch (`conda`, `pip`, source): pip
- If compiling from source, the output of `torch.__config__.show()`:
- Running environment of LightningApp (e.g. local, cloud): n/a
- Any other relevant information: `rich==12.5.1`

### Additional context
Investigating!

https://github.com/Lightning-AI/lightning/pulls?q=is%3Apr+label%3A%22progress+bar%3A+rich%22+milestone%3A%22pl%3A1.6%22+

cc @tchaton @rohitgr7 @kaushikb11 @borda @akihironitta

### PR DESCRIPTION
**Pull request by akihironitta**
## What does this PR do?
Tries to fix #13937 by partially reverting #10497 to revive `auto_refresh=True`.

<details><summary>code for benchmark</summary>

```python
import torch
from torch.utils.data import DataLoader, Dataset
from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.callbacks import RichProgressBar

class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len

class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return loss

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)

def main():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    model = BoringModel()
    trainer = Trainer(
        max_epochs=100,
        enable_model_summary=False,
        enable_checkpointing=False,
        logger=False,
        benchmark=False,  # True by default in 1.6.{0-3}.
        callbacks=RichProgressBar(),
    )
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)

if __name__ == "__main__":
    main()
```

</details>

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by akihironitta** [comment]
(Still wip but) Following https://github.com/Textualize/rich/discussions/2432, it now uses `auto_refresh=True` in this PR, and I see a significant improvement locally:
```
time: 4.771114709   # bugfix/rich-auto-refresh (4ca7033693f0bef01ac724ca83e4c6c4c9550b42)
time: 12.920467208  # master (b84c03f3a6c5327814c6aaab5634e5b5d46919d8)
```

<details><summary>code</summary>

```python
from time import monotonic
import torch
from torch.utils.data import DataLoader, Dataset
from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.callbacks import RichProgressBar

class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len

class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def configure_optimizers(self):
        optimizer = torch.optim.SGD(self.layer.parameters(), lr=0.1)
        lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1)
        return [optimizer], [lr_scheduler]

def main():
    train_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    val_data = DataLoader(RandomDataset(32, 64), batch_size=2)
    model = BoringModel()
    trainer = Trainer(
        max_epochs=100,
        enable_model_summary=False,
        enable_checkpointing=False,
        logger=False,
        benchmark=False,  # True by default in 1.6.{0-3}.
        callbacks=RichProgressBar(),
    )
    t0 = monotonic()
    trainer.fit(model, train_dataloaders=train_data, val_dataloaders=val_data)
    print("time:", monotonic() - t0)

if __name__ == "__main__":
    main()
```

**Comment by akihironitta** [comment]
⚠️ Not ready ⚠️ Marked it as ready just to trigger CI.

**Comment by justusschock** [comment]
@akihironitta How is it going here? Just a friendly reminder :)

**Comment by PKizzle** [comment]
May I ask why this PR has been closed?
